wellcomecollection/wellcomecollection.org
# Digital guides QR logic; support legacy and new

Relates to #11039 

## What does this change?

- Moved the redirection logic to `content/webapp/utils/digital-guides.ts` as it's now used in two different pages. I've also modified it's logic, and I can understand that's tougher to see now. For the old redirections, it's the same thing, I've just slightly tweaked the conditional. I'm hoping it's readable if you have a read through, but let me know if I can help otherwise!
- `isValidType` type guard name changed to `isValidExhibitionGuideType` just for clarity, nothing else was changed.
- Added some type checks for `stopNumber` and `userPreferenceGuideType` to check they are valid, hopefully to avoid some weird bugs that would very much be edge cases.
- This supports the current/old QR code links as well as the upcoming/new ones.

I've left a lot of comments throughout the code so I'm hoping it's clear what's going on; let me know if not.

## How to test

- "Exhibition Guides redesign/restructure work" toggle is ON
- Run locally with the [Prismic endpoint](https://github.com/wellcomecollection/wellcomecollection.org/blob/main/common/services/prismic/fetch/index.ts#L32) set to `wellcomecollection-stage`.

Then:
1. Test all user journeys listed in the [Notion Digital Guides page](https://www.notion.so/wellcometrust/Digital-Guides-948adba1d0ff4bf183242d311f69b55a?pvs=4#59b5dab8495d402993f5a2ab10b40c9e). You'll want to make sure you have the query param `usingQRCode=true` in the URL, but also you'll want to test it doesn't do anything if it's not there.
2. Test that the Jason links work just as well as before ([here they are](https://wellcomecloud.sharepoint.com/:x:/r/sites/wc2/DE/Editorial/_layouts/15/doc2.aspx?sourcedoc=%7B4A6DB307-D923-46F1-964F-201B4B84E64F%7D&file=URLinatedURLs.xlsx&action=default&mobileredirect=true)).

Other things we should check:
- Are the other query params kept (because we need them for tracking)
- If the `WC_userPreferenceGuideType` doesn't have a valid value, does it break things?
- Is the behaviour without the `egWork` toggle affected?


## How can we measure success?

There are no irritants for our users when it gets released, both on the legacy exhibitions and the new ones.

## Have we considered potential risks?

Jason exhibition is affected if we got something wrong, but hopefully we'll have tested it all well enough for it not to happen.

## TODO before merge
- [x] Add e2e tests, potentially in a different child PR (see #11121)
